matchzoo/generators/pair_generator.py

```diff
@@ -0,0 +1,140 @@
+        >>> right.set_index('id_right', inplace=True)
+        >>> input = datapack.DataPack(relation=relation,
+        ...                           left=left,
+        ...                           right=right
+        ... )
+        >>> from matchzoo.generators import PairGenerator
+        >>> generator = PairGenerator(input, 1, 1, 1, 'train', True)
+        >>> assert generator
+        >>> len(generator)
+        2
+        >>> x, y = generator[0]
+        >>> assert x is not None
+        >>> assert y is not None
+
+    """
+
+    def __init__(
+        self,
+        inputs: datapack.DataPack,
+        num_neg: int=1,
+        dup_time: int=4,
+        batch_size: int=32,
+        stage: str='train',
+        shuffle: bool=True
+    ):
+        """Construct the pair generator.
+
+        :param inputs: the output generated by :class:`DataPack`.
+        :param num_neg: the number of negative samples associated with each
+            positive sample.
+        :param dup_time: the number of duplicates for each positive sample.
+        :param batch_size: number of instances in a batch.
+        :param shuffle: whether to shuffle the instances while generating a
+            batch.
+        """
+        self._num_neg = num_neg
+        self._dup_time = dup_time
+        self._left = inputs.left
+        self._right = inputs.right
+        num_rel, self._relation = self.transform_relation(inputs.relation)
+        super().__init__(batch_size, num_rel, stage, shuffle)
+
+    def transform_relation(self, relations: pd.DataFrame) -> pd.DataFrame:
+        """Obtain the transformed data from :class:`DataPack`.
+
+        Note here, label is required to make pairs.
+
+        :param relations: An instance of :class:`DataPack` to be transformed.
+        :return: the output of all the transformed relations.
+        """
+        if 'label' not in relations.columns:
+            raise ValueError(f"label is required from {relations} \
+                             to generate pairs.")
+        # Note here the main id is set to be the id_left
+        pairs = []
+        num_pairs = 0
+        for idx, group in relations.sort_values('label', ascending=False).\
+                groupby('id_left'):
+            labels = group.label.unique()
+            for label in labels:
+                pos_samples = group[group.label == label]
+                pos_samples = pd.concat([pos_samples] * self._dup_time)
+                neg_samples = group[group.label < label]
+                for _, pos_sample in pos_samples.iterrows():
+                    pos_sample = pd.DataFrame([pos_sample])
+                    if len(neg_samples) >= self._num_neg:
+                        neg_sample = neg_samples.sample(self._num_neg,
+                                                        replace=False)
+                        pairs.extend((pos_sample, neg_sample))
+                        num_pairs += 1
+        return num_pairs, pd.concat(pairs, ignore_index=True)
```

Comment:
why `num_pairs` and add 1 per iteration? why not `return len(pairs)`?

Reply:
Here, len(pairs) not equal to num_pairs. For case `num_neg = 2`, then, `pairs = [pos0, neg00, neg01, pos1, neg10, neg11, ...]`. Thus, `num_pairs = len(pairs) / (num_neg + 1)`.